packages/avro-ts/src/helpers.ts

```diff
@@ -32,19 +32,46 @@ export const namedType = (
 ): Document<ts.TypeNode, Context> => {
   const name = convertName(firstUpperCase(schema.name));
   const namespaceName = namespace ? convertName(namespace) : undefined;
+
   const fullName = namespaceName ? [namespaceName, name] : name;
+  const fieldName = `${name}Name`;
+  const schemaName = `${namespace}.${fieldName}`;
 
-  const contextWithRef = namespace
-    ? withIdentifier(
+  let contextWithRef: Context;
+
+  if (namespace) {
+    // If there is already a ref with the same name as our "named type", it means there is already
+    // a type with the same name and we're about to have a naming collision. To avoid this, we
+    // use the fully qualified name instead.
+    if (context.refs && schemaName in context.refs) {
+      contextWithRef = withIdentifier(
         context,
         Node.Const({
-          name: `${name}Name`,
+          name: `${namespaceName}${fieldName}`,
           isExport: true,
           value: `${namespace}.${schema.name}`,
         }),
         namespaceName,
-      )
-    : context;
+      );
+    } else {
+      contextWithRef = withIdentifier(
+        context,
+        Node.Const({
+          name: fieldName,
+          isExport: true,
+          value: `${namespace}.${schema.name}`,
+        }),
+        namespaceName,
+      );
+    }
+  } else {
+    contextWithRef = context;
+  }
```

Comment:
I know this is a bit silly, but I'd prefer to avoid "let" if possible. It makes absolutely no difference in this specific case, but it allows the code to be easily "dirtied" later, as it will make it trivial to insert logic that modifies the `contextWithRef` later and I'd wish to avoid that style of code.

Also no need for the result variable too.

```typescript
  const value = `${namespace}.${schema.name}`;

  const contextWithRef = namespace
    ? /**
       * If there is already a ref with the same name as our "named type", it means there is already
       * a type with the same name and we're about to have a naming collision. To avoid this, we
       * use the fully qualified name instead.
       */
      context.refs && schemaName in context.refs
      ? withIdentifier(
          context,
          Node.Const({ name: `${namespaceName}${fieldName}`, isExport: true, value }),
          namespaceName,
        )
      : withIdentifier(
          context,
          Node.Const({ name: fieldName, isExport: true, value }),
          namespaceName,
        )
    : context;

  return document(withIdentifier(contextWithRef, type, namespaceName), Type.Referance(fullName));
```

something like this looks much nicer to my eyes :-D

Reply:
That works for me! I'll update according and push the changes up here in a bit.